### PR TITLE
issue-431:Convert Pytest-CVP IP Peers Established test case to Vane

### PR DESCRIPTION
**Pull request by rohitsinghgslab**
# Please include a summary of the changes

test_definition.yaml: Updated test def with testcase NRFU4.2
test_definition.yaml.j2: Updated J2 with testcase NRFU4.2
test_bgp_peers_state.py: Added python file for testcase NRFU4.2

# Any specific logic/part of code you need extra attention on

Validating BGP ipv4 peers are established.

# Include the Issue number and link
#431 

# List/Attach any dependencies/past issues that are required for this change/provide context

# Type of change

- - [ ] Bug fix (non-breaking change which fixes an issue)
- - [ ] New feature (non-breaking change which adds functionality)
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- - [ ] This change requires a documentation update
- - [X] Other (NRFU testcases)

# Effort required on reviewers end

- - [ ] Easy
- - [X] Medium
- - [ ] Hard 

# How Has This Been Tested?
**Note: Tested on ACT, only failed case is tested on hard - coded.**
Unit tests and reports :

Pass report: All BGP IPv4 peers are in established state.(J2)
[report_2309151741.docx](https://github.com/aristanetworks/vane/files/12619682/report_2309151741.docx)

Fail report: BGP peers are not in established state.(J2)
[report_2309051624.docx](https://github.com/aristanetworks/vane/files/12522742/report_2309051624.docx)
[report_2309051635.docx](https://github.com/aristanetworks/vane/files/12522744/report_2309051635.docx)

Skipped: No IP BGP configured (Tested on ACT with DCBBE1 dut.).(J2)
[report_2309151818.docx](https://github.com/aristanetworks/vane/files/12619664/report_2309151818.docx)

Assert: If BGP peers are not found for any VRFs..
[report_2309151911.docx](https://github.com/aristanetworks/vane/files/12619942/report_2309151911.docx)

HTML reports(J2)
[Vane_PS_431.zip](https://github.com/aristanetworks/vane/files/12619946/Vane_PS_431.zip)

Passed_report_without_J2
[passed_without_J2.zip](https://github.com/aristanetworks/vane/files/12522867/passed_without_J2.zip)

[reports_after_resolving_conflict.zip](https://github.com/aristanetworks/vane/files/12535556/reports_after_resolving_conflict.zip)

Pass report after adding logging and F-string changes
[Pass_report_after_logging_changes.zip](https://github.com/aristanetworks/vane/files/12607106/Pass_report_after_logging_changes.zip)

## New feature

 
    
# CI pipeline result

- - [ ] Pass
- - [ ] Fail
  
  If it fails, explain the reason and whether or not we should ignore the failure
  
# Verify Documentation Update

  
    
# Additional comments

### CODE REVIEW

File: nrfu_tests/master_def.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+common_data:
+  report_style: modern # If report type is set to modern, then it will update steps, assumptions, and external systems in the docx report.
+testcase_data:
+  NRFU4.2:
+    criteria: names
+    filter:
+      - BLFE1
```
Comment (hsisalgslab) [inline nrfu_tests/master_def.yaml:7]: update as per latest placeholder MR

File: nrfu_tests/test_bgp_peers_state.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            dut(dict): details related to a particular DUT
+            tests_definitions(dict): test suite and test case parameters.
+        """
+        tops = tests_tools.TestOps(tests_definitions, TEST_SUITE, dut)
+        tops.expected_output = {"bgp_peers": {}}
+        tops.actual_output = {"bgp_peers": {}}
+        self.output = ""
+
+        # forming output message if test result is passed
+        tops.output_msg = "All BGP peers are in established state."
```
Comment (hsisalgslab) [inline nrfu_tests/test_bgp_peers_state.py:42]:
```suggestion
        tops.output_msg = "All BGP IPv4 peers are in established state."
```

File: nrfu_tests/test_bgp_peers_state.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+        try:
+            """
+            TS: Running `show ip bgp summary vrf all` command and verifying
+            state of BGP peers.
```
Comment (hsisalgslab) [inline nrfu_tests/test_bgp_peers_state.py:47]:
```suggestion
            state of BGP IPv4 peers.
```

File: nrfu_tests/test_bgp_peers_state.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            output = tops.run_show_cmds([tops.show_cmd])
+            logger.info(
+                "On device %s, output of %s command is:\n%s\n",
+                tops.dut_name,
+                tops.show_cmd,
+                output,
+            )
+            self.output = f"\nOutput of {tops.show_cmd} command is:\n{output}\n"
+            bgp_peers = output[0]["result"]["vrfs"]
+            assert bgp_peers, "BGP peers details are not found."
```
Comment (hsisalgslab) [inline nrfu_tests/test_bgp_peers_state.py:58]:
```suggestion
            assert bgp_peers, "BGP peer details are not found."
```

File: nrfu_tests/test_bgp_peers_state.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                    tops.actual_output["bgp_peers"].update({peer: peer_state})
+                    tops.expected_output["bgp_peers"].update({peer: "Established"})
+
+            # Forming output message if test result is fail.
+            if tops.actual_output != tops.expected_output:
+                tops.output_msg = "\n"
+                for interface, peer_state in tops.expected_output["bgp_peers"].items():
+                    if peer_state != tops.actual_output["bgp_peers"].get(interface):
+                        tops.output_msg += (
+                            f"For the BGP peer '{interface}':\nExpected peer state is {peer_state},"
```
Comment (hsisalgslab) [inline nrfu_tests/test_bgp_peers_state.py:73]:
```suggestion
                            f"For the BGP peer '{interface}':\nExpected peer state is '{peer_state}',"
```

File: nrfu_tests/test_bgp_peers_state.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                            " however actual found as"
+                            f" {tops.actual_output['bgp_peers'].get(interface)}.\n\n"
```
Comment (hsisalgslab) [inline nrfu_tests/test_bgp_peers_state.py:75]:
```suggestion
                            f" '{tops.actual_output['bgp_peers'].get(interface)}'.\n\n"
```

**Review by hsisalgslab — APPROVED** [review]

File: nrfu_tests/master_def.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+global_dut_filter: # This stanza is optional. If it is not provided by users than all test cases should be run on each dut.
+  criteria: names
+  filter:
```
Comment (geetanjalimanegslab) [inline nrfu_tests/master_def.yaml:3]: As discussed  no need of this file.

File: nrfu_tests/test_bgp_peers_state.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
+# Copyright (c) 2023 Arista Networks, Inc.  All rights reserved.
+# Arista Networks, Inc. Confidential and Proprietary.
+
+"""
+Testcase for verification of BGP IPv4 peers are established.
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_bgp_peers_state.py:5]: Update the same at all other places as well.
```suggestion
Testcase for verification of BGP IPv4 peers state.
```

File: nrfu_tests/test_bgp_peers_state.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            state of BGP IPv4 peers.
+            """
+            output = tops.run_show_cmds([tops.show_cmd])
+            logger.info(
+                "On device %s, output of %s command is:\n%s\n",
+                tops.dut_name,
+                tops.show_cmd,
+                output,
+            )
+            self.output = f"\nOutput of {tops.show_cmd} command is:\n{output}\n"
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_bgp_peers_state.py:56]:
```suggestion
            self.output = f"Output of {tops.show_cmd} command is:\n{output}\n"
```

File: nrfu_tests/test_bgp_peers_state.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            bgp_peers = output[0]["result"]["vrfs"]
+            assert bgp_peers, "BGP peer details are not found."
+
+            # Collecting actual and expected output.
+            for vrf in bgp_peers:
+                for peer in bgp_peers.get(vrf).get("peers"):
+                    peer_state = bgp_peers.get(vrf).get("peers").get(peer).get("peerState")
+                    tops.actual_output["bgp_peers"].update({peer: peer_state})
+                    tops.expected_output["bgp_peers"].update({peer: "Established"})
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_bgp_peers_state.py:65]: Add this line  tops.expected_output["bgp_peers"].update({peer: "Established"}) just after the 2nd for loop on line number 63

File: nrfu_tests/test_bgp_peers_state.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+            # Forming output message if test result is fail.
+            if tops.actual_output != tops.expected_output:
+                tops.output_msg = "\n"
+                for interface, peer_state in tops.expected_output["bgp_peers"].items():
+                    if peer_state != tops.actual_output["bgp_peers"].get(interface):
+                        tops.output_msg += (
+                            f"For the BGP peer {interface}:\nExpected peer state is '{peer_state}',"
+                            " however actual found as"
+                            f" '{tops.actual_output['bgp_peers'].get(interface)}'.\n\n"
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_bgp_peers_state.py:75]: can we use a variable to store the   `tops.actual_output['bgp_peers'].get(interface)`

File: nrfu_tests/test_definition.yaml.j2 [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+  testcases:
+    - name: test_bgp_ipv4_peers_state
+      test_id: NRFU4.2
+      description: Testcase for verification of BGP IPv4 peers are established.
+      show_cmd: show ip bgp summary vrf all
+      expected_output: null
+      test_criteria: All BGP IPv4 peers should be in established state.
+      report_style: modern
+      criteria: {{ global_dut_filter.criteria }}
+      filter: {{ global_dut_filter.filter }}
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_definition.yaml.j2:11]: please confirm that the new line is added at the end of the file.

File: nrfu_tests/test_bgp_peers_state.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+                    tops.expected_output["bgp_peers"].update({peer: "Established"})
+
+            # Forming output message if test result is fail.
+            if tops.actual_output != tops.expected_output:
+                tops.output_msg = "\n"
+                for interface, peer_state in tops.expected_output["bgp_peers"].items():
+                    if peer_state != tops.actual_output["bgp_peers"].get(interface):
+                        tops.output_msg += (
+                            f"For the BGP peer {interface}:\nExpected peer state is '{peer_state}',"
+                            " however actual found as"
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_bgp_peers_state.py:74]:
```suggestion
                            " however actual is found as"
```

File: nrfu_tests/test_bgp_peers_state.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+@pytest.mark.nrfu_test
+@pytest.mark.routing
+class BgpIpPeersStatusTests:
+    """
+    Testcase for verification of BGP IPv4 peers are established.
+    """
+
+    dut_parameters = tests_tools.parametrize_duts(TEST_SUITE, test_defs, dut_objs)
+    test_duts = dut_parameters["test_bgp_ipv4_peers_state"]["duts"]
+    test_ids = dut_parameters["test_bgp_ipv4_peers_state"]["ids"]
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_bgp_peers_state.py:26]: Add report for assertion failures.

**Review by geetanjalimanegslab — CHANGES_REQUESTED** [review]

File: nrfu_tests/test_definition.yaml.j2 [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+    - name: test_bgp_ipv4_peers_state
+      test_id: NRFU4.2
+      description: Testcase for verification of BGP IPv4 peers state.
+      show_cmd: show ip bgp summary vrf all
+      expected_output: null
+      test_criteria: All BGP IPv4 peers should be in established state.
+      report_style: modern
+      criteria: {{ global_dut_filter.criteria }}
+      filter: {{ global_dut_filter.filter }}
+
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_definition.yaml.j2:12]: the extra line is added.

**Review by geetanjalimanegslab — CHANGES_REQUESTED** [review]

File: nrfu_tests/test_bgp_peers_state.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+        # forming output message if test result is passed
+        tops.output_msg = "All BGP IPv4 peers are in established state."
+
+        try:
+            """
+            TS: Running `show ip bgp summary vrf all` command and verifying
+            state of BGP IPv4 peers.
+            """
+            output = tops.run_show_cmds([tops.show_cmd])
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_bgp_peers_state.py:49]: Why do we need to run the show cmd again? as we are collecting the output of same in the test def yaml

**Review by geetanjalimanegslab — CHANGES_REQUESTED** [review]

File: nrfu_tests/test_bgp_peers_state.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
+# Copyright (c) 2023 Arista Networks, Inc.  All rights reserved.
+# Arista Networks, Inc. Confidential and Proprietary.
+
+"""
+Testcase for verification of BGP IPv4 peers state.
+"""
+
+import pytest
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_bgp_peers_state.py:8]: please mark PR as draft

Comment (rohitsinghgslab) [reply nrfu_tests/test_bgp_peers_state.py:8]: marked

**Review by geetanjalimanegslab — APPROVED** [review]

File: nrfu_tests/test_bgp_peers_state.py [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        tops.actual_output = {"bgp_peers": {}}
+        self.output = ""
+
+        # forming output message if test result is passed
+        tops.output_msg = "All BGP IPv4 peers are in established state."
+
+        try:
+            """
+            TS: Running `show ip bgp summary vrf all` command and verifying
+            state of BGP IPv4 peers.
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_bgp_peers_state.py:47]:
```suggestion
            TS: Running `show ip bgp summary vrf all` command on dut and verifying
            state for all BGP IPv4 peers.
```

File: nrfu_tests/test_bgp_peers_state.py [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+            output = dut["output"][tops.show_cmd]["json"]
+            logger.info(
+                "On device %s, output of %s command is:\n%s\n",
+                tops.dut_name,
+                tops.show_cmd,
+                output,
+            )
+            self.output = f"Output of {tops.show_cmd} command is:\n{output}\n"
+            bgp_peers = output.get("vrfs")
+            if not bgp_peers:
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_bgp_peers_state.py:58]: Testcase should skip if peers are not found in any VRF
image

File: nrfu_tests/test_bgp_peers_state.py [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+                    actual_peer_state = tops.actual_output["bgp_peers"].get(interface).get("state")
+                    expected_peer_state = peer_state.get("state")
+                    if expected_peer_state != actual_peer_state:
+                        tops.output_msg += (
+                            f"For the BGP peer {interface}:\nExpected peer state is"
+                            f" '{expected_peer_state}', however actual is found as"
+                            f" '{actual_peer_state}'.\n\n"
+                        )
+
+        except (AssertionError, AttributeError, LookupError, EapiError) as excep:
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_bgp_peers_state.py:81]: there is no assert so remove AssertionError exception

File: nrfu_tests/test_bgp_peers_state.py [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+
+            # Forming output message if test result is fail.
+            if tops.actual_output != tops.expected_output:
+                tops.output_msg = "\n"
+                for interface, peer_state in tops.expected_output["bgp_peers"].items():
+                    actual_peer_state = tops.actual_output["bgp_peers"].get(interface).get("state")
+                    expected_peer_state = peer_state.get("state")
+                    if expected_peer_state != actual_peer_state:
+                        tops.output_msg += (
+                            f"For the BGP peer {interface}:\nExpected peer state is"
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_bgp_peers_state.py:76]: Dut output and observation is not matching. You can change the config get the proper dut output and observation, No need to use hardcode output
image

File: nrfu_tests/test_bgp_peers_state.py [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+                tops.show_cmd,
+                output,
+            )
+            self.output = f"Output of {tops.show_cmd} command is:\n{output}\n"
+            bgp_peers = output.get("vrfs")
+
+            # Skipping, if no IP BGP peers configured.
+            for vrf in bgp_peers:
+                if bgp_peers.get(vrf).get("peers"):
+                    bgp_peers_found = True
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_bgp_peers_state.py:63]: if you get True in first VRF only then no need to check other VRF so add a break

File: nrfu_tests/test_bgp_peers_state.py [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+            if not bgp_peers_found:
+                pytest.skip(f"For {tops.dut_name}, no IP BGP peers configured.")
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_bgp_peers_state.py:65]:
```suggestion
                pytest.skip(f"For {tops.dut_name}, BGP peers are not found hence skipped the testcase.")
```

File: nrfu_tests/test_bgp_peers_state.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+                    tops.actual_output["bgp_peers"].update({peer: {"state": peer_state}})
+
+            # Forming output message if test result is fail.
+            if tops.actual_output != tops.expected_output:
+                tops.output_msg = "\n"
+                for interface, peer_state in tops.expected_output["bgp_peers"].items():
+                    actual_peer_state = tops.actual_output["bgp_peers"].get(interface).get("state")
+                    expected_peer_state = peer_state.get("state")
+                    if expected_peer_state != actual_peer_state:
+                        tops.output_msg += (
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_bgp_peers_state.py:81]: Form the observation as below. Suggested format was from Kevin
image

Comment (MaheshGSLAB) [reply nrfu_tests/test_bgp_peers_state.py:81]: BGP neighbors state for following peers is not Established:
Neighbor 172.31.255.0 state is idel
Neighbor 172.31.255.2 state is active

**Review by MaheshGSLAB — CHANGES_REQUESTED** [review]

File: nrfu_tests/test_definition.yaml [hunk 25]
```diff
@@ -0,0 +1,12 @@
+- name: nrfu_tests
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_definition.yaml:1]: resolve the conflicts

File: nrfu_tests/test_bgp_peers_state.py [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+            )
+            self.output = f"Output of {tops.show_cmd} command is:\n{output}\n"
+            bgp_peers = output.get("vrfs")
+
+            # Skipping, if no IP BGP peers configured.
+            for vrf in bgp_peers:
+                if bgp_peers.get(vrf).get("peers"):
+                    bgp_peers_found = True
+                    break
+            if not bgp_peers_found:
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_bgp_peers_state.py:65]: awaiting update from Kevin about testcase should fail or skip in case no peers found

**Review by MaheshGSLAB — CHANGES_REQUESTED** [review]

File: nrfu_tests/test_definition.yaml (hunk 25 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_definition.yaml]: resolved it. FYI @MaheshGSLAB

File: nrfu_tests/test_bgp_peers_state.py (hunk 26 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_bgp_peers_state.py:65]: okay

**Review by MaheshGSLAB — APPROVED** [review]

**Review by kgrozis-arista — APPROVED** [review]

**Pull request merged**